File: 0036-valid-sudoku/0036-valid-sudoku.rb
# @param {Character[][]} board
# @return {Boolean}
def is_valid_sudoku(board)
    rows = {}
    cols = {}
    squares = {}
    
    board.each_with_index do |row, r|
       row.each_with_index do |num, c|
           next if num == '.'
           
           s = "#{r / 3}#{c / 3}"
           rows[r] ||= Set.new
           cols[c] ||= Set.new
           squares[s] ||= Set.new
           
           if rows[r].include?(num) || cols[c].include?(num) || squares[s].include?(num)
               return false
           end
           
           rows[r].add(num)
           cols[c].add(num)
           squares[s].add(num)
       end
    end
    
    true
end

File: 0036-valid-sudoku/0036-valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
const isValidSudoku = function(board) {
    const rows = {},
          cols = {},
          boxes = {};
    
    const invalid = (r, c, b, v) => {
        return rows[r].has(v)|| 
               cols[c].has(v) ||
               boxes[b].has(v);
    }
    
    for (let r = 0; r < 9; r++) {
        for (let c = 0; c < 9; c++) {
            const b = `${~~(r/3)}${~~(c / 3)}`,
                  val = board[r][c];
            
            if (val === '.') continue;
            
            rows[r] ||= new Set();
            cols[c] ||= new Set();
            boxes[b] ||= new Set();
            
            if (invalid(r, c, b, val)) return false;
            
            rows[r].add(val);
            cols[c].add(val);
            boxes[b].add(val);
        }
    }
    
    return true;
};